src/Microsoft.Graph.Core/Requests/RetryHandler.cs

```diff
@@ -0,0 +1,193 @@
+        /// <param name="response">The <see cref="HttpResponseMessage"/>returned.</param>
+        /// <returns></returns>
+        public bool IsRetry(HttpResponseMessage response)
+        {
+            if ((response.StatusCode == HttpStatusCode.ServiceUnavailable ||
+                response.StatusCode == (HttpStatusCode)429))
+            {
+                return true;
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Check the HTTP request's content to determine whether it can be retried or not.
+        /// </summary>
+        /// <param name="request">The <see cref="HttpRequestMessage"/>needs to be sent.</param>
+        /// <returns></returns>
+        private bool IsBuffed(HttpRequestMessage request)
+        {
+            HttpContent content = request.Content;
+
+            if ((request.Method == HttpMethod.Put || request.Method == HttpMethod.Post) 
+                && (content == null || content.Headers.ContentLength == null || (int)content.Headers.ContentLength == -1))
+            {
+                return false;
+            }
+            return true;
+           
+        }
+
+        /// <summary>
+        /// Update Retry-Attempt header in the HTTP request
+        /// </summary>
+        /// <param name="request">The <see cref="HttpRequestMessage"/>needs to be sent.</param>
+        /// <param name="retry_count">Retry times</param>
+        private void AddOrUpdateRetryAttempt(HttpRequestMessage request, int retry_count)
+        {
+            if (request.Headers.Contains(RETRY_ATTEMPT))
+            {
+                request.Headers.Remove(RETRY_ATTEMPT);
+            }
+            request.Headers.Add(RETRY_ATTEMPT, retry_count.ToString());
+        }
+
+        /// <summary>
+        /// Delay task operation based on Retry-After header in the response or exponential backoff
+        /// </summary>
+        /// <param name="response">The <see cref="HttpResponseMessage"/>returned.</param>
+        /// <param name="retry_count">The retry counts</param>
+        /// <returns>The <see cref="Task"/> for delay operation.</returns>
+        public Task Delay(HttpResponseMessage response, int retry_count)
+        {
+            
+            TimeSpan delay = TimeSpan.FromMilliseconds(0);
+            HttpHeaders headers = response.Headers;
+            if (headers.TryGetValues(RETRY_AFTER, out IEnumerable<string> values))
+            {
+                string retry_after = values.First();
+                if (Int32.TryParse(retry_after, out int delay_seconds))
+                {
+                    delay = TimeSpan.FromSeconds(delay_seconds);
+                }
+            }
+            else
+            {
+                if (retry_count < 31)
+                {
+                    m_pow = m_pow << 1; // m_pow = Pow(2, m_retries - 1)
+                }
+                int delay_time = Math.Min(DELAY_MILLISECONDES * (m_pow - 1) / 2,
+                    MAX_DELAY_MILLISECONDS);
```

Comment:
MAX_DELAY_MILLISECONDS - the value is 10000. Is this value arbitrary, did we get it from AGS, or did we get it from other source? What's the thought around 10 seconds here?